### DISCUSSION TITLE
Jackrouter gives a better performance than direct connection to the audio device

### DISCUSSION THREAD

**Discussion by birkeeper** [jamulussoftware/jamulus#1441] [Technical Support & Help]
I use a Windows (64-bit) computer with a M-Audio M-track solo USB audio device. I have installed the ASIO driver provided by the manufacturer. I also have JackCtl and the JackRouter ASIO driver installed.
**When I connect Jamulus directly to the USB audio device using the M-Audio ASIO driver, it performs worse then when I connect it through the JackRouter ASIO driver. Does anyone have an explanation?**

**Background info:**
_General settings:_
Jamulus server running at localhost --> ping time < 1 ms
local jitter buffer size: 4
server jitter buffer size: 4
audio channels: mono-in/stereo-out
audio quality: normal

_Setup M-audio ASIO driver_
Selected audio device: M-audio ASIO driver
buffer delay: 64 frames
M-audio ASIO driver buffer settings: 64 frames
--> this results in a total delay of approximately **31 ms** and regular local jitter buffer errors

_Setup JackRouter ASIO driver_
Selected audio device: JackRouter
In JackCtl the M-audio ASIO driver is selected and set to 48 kHz and 64 bit frames
--> this results in a total delay of approximately **27 ms**. local jitter buffer errors occur less often then in the previous setup

**Comment by dcorson-ticino-com** [comment]
Maybe M-Audio has an updated driver with better performance ?
Unfortunately at the price level of the M-Track Solo I don't think the manufacturers spend much time optimizing the drivers.
You could also try ASIO4ALL to see if it is better.

**Reply by birkeeper** [reply]
I tried ASIO4ALL, but this performs worse. 
What I do not understand is that JackCtl communicates with the same M-audio ASIO driver as Jamulus, but that the more complex setup performs better. What is the difference between the way Jamulus and JackCtl communicate with the M-audio ASIO driver?

**Reply by dcorson-ticino-com** [reply]
This may also depend on how Jamulus approximates the total delay.
I do not know the details, but Jamulus does not measure the total delay, it is approximated using several parameters.
It would be necessary to set up an oscilloscope etc. showing the waveform at the input and at the output to really measure the total delay.

**Reply by birkeeper** [reply]
So are you suggesting that JackCtl adds a buffer that removes the jitter of the audio device and that this extra buffer delay is not detected by Jamulus?

**Reply by dcorson-ticino-com** [reply]
No, I am suggesting that Jamulus' approximation of the total delay may be somewhat simplistic and may be distorting the true values.

EDIT----------------------
The estimation routine is in client.cpp,  EstimatedOverallDelay ( iPingTime );
It depends on several different buffers (i.e. internal to Jamulus) and also on what the sound card returns as its delay (external to Jamulus).  If the sound card doesn't return a value a value is assumed.  That may be where the difference occurs.
But that is just supposition and looking for a possible cause.  Without testing there is no way to really know.

**Reply by birkeeper** [reply]
Okay, so it could be that JackCtl reports its own delay to Jamulus and that the delay is probably not better than in the direct setup.  I guess that the fact that there are less cracks and pops (less buffer jitter) with the JackCtl setup may be due to buffers in JackCtl.
Thank you for looking into the issue.

**Reply by birkeeper** [reply]
Update: used the ASIO4ALL v2.15 Beta 1 instead of M-audio driver. The result is a reduced delay of approximately 5 ms and less cracks and pops.

**Comment by birkeeper** [comment]
I have added my experiences with the M-track Solo to the Wiki page. I consider this question as being answered. Thank you.
